### DISCUSSION TITLE
standardnotes: JavaScript error & profile autodetect (LXQt)

### DISCUSSION THREAD

**Discussion by hotcapy** [netblue30/firejail#5758] [Q&A] [answered]
Hi!

I'm trying to run **[Standard Notes AppImage](https://github.com/standardnotes/app)** with **Firejail**, but getting errors with both `default` and `standardnotes-desktop` global profiles. Results are same no matter which directory I choose for placing AppImage: `/opt` or `/home/user/Software/Standard-Notes`.

1) Error with `default` profile related to Firejail itself:

```
[user@host ~]$ firejail --appimage /opt/standard-notes-3.150.7-linux-x86_64.AppImage
Reading profile /etc/firejail/default.profile
Reading profile /etc/firejail/disable-common.inc
Reading profile /etc/firejail/disable-programs.inc

** Note: you can use --noprofile to disable default.profile **

Parent pid 4887, child pid 4889

** Warning: dropping all Linux capabilities and setting NO_NEW_PRIVS prctl **

Warning: NVIDIA card detected, nogroups command ignored
Mounting appimage type 2
Warning: /sbin directory link was not blacklisted
Warning: /usr/sbin directory link was not blacklisted
Warning: NVIDIA card detected, nogroups command ignored
Child process initialized in 60.41 ms
[6:0327/001832.506957:FATAL:setuid_sandbox_host.cc(157)] The SUID sandbox helper binary was found, but is not configured correctly. Rather than run without sandboxing I'm aborting now. You need to make sure that /run/firejail/appimage/chrome-sandbox is owned by root and has mode 4755.

Parent is shutting down, bye...
AppImage detached
```

I have mentioned `/run/firejail/appimage` directory, but it's empty - no `chrome-sandbox` file or folder inside it.

2) With `standardnotes-desktop` profile error changes into this:

```
[user@host ~]$ firejail --profile=standardnotes-desktop --appimage /opt/standard-notes-3.150.7-linux-x86_64.AppImage
Reading profile /etc/firejail/standardnotes-desktop.profile
Reading profile /home/user/.config/firejail/standardnotes-desktop.local
Reading profile /etc/firejail/disable-common.inc
Reading profile /etc/firejail/disable-devel.inc
Reading profile /etc/firejail/disable-exec.inc
Reading profile /etc/firejail/disable-interpreters.inc
Reading profile /etc/firejail/disable-programs.inc
Reading profile /etc/firejail/whitelist-var-common.inc
Seccomp list in: !chroot, check list: @default-keep, prelist: unknown,
Parent pid 4538, child pid 4539

** Warning: dropping all Linux capabilities and setting NO_NEW_PRIVS prctl **

Warning: NVIDIA card detected, nogroups command ignored
Mounting appimage type 2
Warning: An abstract unix socket for session D-BUS might still be available. Use --net or remove unix from --protocol set.
Private /etc installed in 36.79 ms
Warning: NVIDIA card detected, nogroups command ignored
Warning: NVIDIA card detected, nogroups command ignored
Warning: /sbin directory link was not blacklisted
Warning: /usr/sbin directory link was not blacklisted
Warning: not remounting /run/user/1000/doc
Seccomp list in: !chroot, check list: @default-keep, prelist: unknown,
Warning: NVIDIA card detected, nogroups command ignored
Warning: NVIDIA card detected, nogroups command ignored
Child process initialized in 127.91 ms
A JavaScript error occurred in the main process
Uncaught Exception:                                                                                                                                                                            
Error: /tmp/.org.chromium.Chromium.DeWXwA: failed to map segment from shared object                                                                                                            
    at process.func [as dlopen] (node:electron/js2c/asar_bundle:2:1822)                                                                                                                        
    at Module._extensions..node (node:internal/modules/cjs/loader:1226:18)                                                                                                                     
    at Object.func [as .node] (node:electron/js2c/asar_bundle:2:2049)                                                                                                                          
    at Module.load (node:internal/modules/cjs/loader:1011:32)                                                                                                                                  
    at Module._load (node:internal/modules/cjs/loader:846:12)                                                                                                                                  
    at f._load (node:electron/js2c/asar_bundle:2:13330)                                                                                                                                        
    at Module.require (node:internal/modules/cjs/loader:1035:19)                                                                                                                               
    at require (node:internal/modules/cjs/helpers:102:18)                                                                                                                                      
    at Object.<anonymous> (/run/firejail/appimage/resources/app.asar/node_modules/keytar/lib/keytar.js:1:14)                                                                                   
    at Module._compile (node:internal/modules/cjs/loader:1141:14)                                                                                                                              
[31:0327/001704.138574:ERROR:bus.cc(399)] Failed to connect to the bus: Failed to connect to socket /run/firejail/mnt/dbus/system: Permission denied
[31:0327/001704.138659:ERROR:bus.cc(399)] Failed to connect to the bus: Failed to connect to socket /run/firejail/mnt/dbus/system: Permission denied
[31:0327/001704.147014:ERROR:bus.cc(399)] Failed to connect to the bus: Failed to connect to socket /run/firejail/mnt/dbus/user: Permission denied
[31:0327/001704.147091:ERROR:bus.cc(399)] Failed to connect to the bus: Failed to connect to socket /run/firejail/mnt/dbus/user: Permission denied
```
You can see here some dbus errors in the bottom of output, and creating `standardnotes-desktop.local` file with ignore options removes them:
```
ignore dbus-user none
ignore dbus-system none
```
However, JavaScript error persists, and app does not start (nothing happens).

3) When executed without any profile (`--noprofile`) everything works as expected.

Is there a way to fix it with `standardnotes-desktop` profile or should I request AppImage specific profile for Standard Notes in related issue here? Thanks in advance! <3

**Comment by rusty-snake** [comment]
You can not use `--appimage` with chrom\* programs.

To try to make it work:
1. Start it with the correct profile.
2. Start it by executing the AI rather than passing `--appimage`.
=> `firejail --profile=standardnotes-desktop /opt/standard-notes-3.150.7-linux-x86_64.AppImage`

**Reply by hotcapy** [reply]
Oh, I see.

1) Unfortunately, something is wrong with **fuse** this time:

```
[user@host ~]$ firejail --profile=standardnotes-desktop /opt/standard-notes-3.150.7-linux-x86_64.AppImage
Reading profile /etc/firejail/standardnotes-desktop.profile                                                                                                                                    
Reading profile /etc/firejail/disable-common.inc                                                                                                                                               
Reading profile /etc/firejail/disable-devel.inc
Reading profile /etc/firejail/disable-exec.inc
Reading profile /etc/firejail/disable-interpreters.inc
Reading profile /etc/firejail/disable-programs.inc
Reading profile /etc/firejail/whitelist-var-common.inc
Seccomp list in: !chroot, check list: @default-keep, prelist: unknown,
Parent pid 8047, child pid 8048
Warning: NVIDIA card detected, nogroups command ignored
Warning: An abstract unix socket for session D-BUS might still be available. Use --net or remove unix from --protocol set.
Private /etc installed in 41.23 ms
Warning: NVIDIA card detected, nogroups command ignored
Warning: NVIDIA card detected, nogroups command ignored
Warning: /sbin directory link was not blacklisted
Warning: /usr/sbin directory link was not blacklisted
Warning: not remounting /run/user/1000/doc
Seccomp list in: !chroot, check list: @default-keep, prelist: unknown,
Warning: NVIDIA card detected, nogroups command ignored
Warning: NVIDIA card detected, nogroups command ignored
Child process initialized in 121.68 ms
fuse: device not found, try 'modprobe fuse' first

Cannot mount AppImage, please check your FUSE setup.
You might still be able to extract the contents of this AppImage 
if you run it with the --appimage-extract option. 
See https://github.com/AppImage/AppImageKit/wiki/FUSE 
for more information
open dir error: No such file or directory

Parent is shutting down, bye...
```

I use custom kernel with **modprobed-db**, but fuse module is listed in my config file & actually loaded by default according to `/proc/modules`.

2) Issue is not AppImage specific for me: I builded **standardnotes-desktop** package from **AUR** and getting same JavaScript error as AppImage before when running with Firejail. Also, like with AppImage, Firejail not detects Standard Notes profile automatically and fallbacks to default one, without explicitly setting it as option. In this package executable is `/usr/bin/standard-notes`.

Anyway, thanks for your answer!

**Comment by ghost** [comment, marked as answer]
> [6:0327/001832.506957:FATAL:setuid_sandbox_host.cc(157)] The SUID sandbox helper binary was found, but is not configured correctly. Rather than run without sandboxing I'm aborting now. You need to make sure that /run/firejail/appimage/chrome-sandbox is owned by root and has mode 4755.

I've seen similar errors with (chrome-based) AppImage files before.

> You can not use --appimage with chrom* programs.

There is in fact a way to make --appimage work via the `--no-sandbox` option. Note that this switch is _not_ relating to the firejail sandbox but to the internal chrome-sandbox mechanism, and because of this it needs to be placed accordingly. For me the below does work:

```
$ firejail --ignore="noexec /tmp" --private-opt=standardnotes --profile=standardnotes-desktop --appimage /opt/standardnotes/standard-notes-3.150.7-linux-x86_64.AppImage --no-sandbox
```

One could create a ~/.config/firejail/standardnotes-desktop.local and place the needed firejail options there:

```console
$ cat ~/.config/firejail/standardnotes-desktop.local
# Firejail profile for standardnotes-desktop
# Persistent local customizations

?HAS_APPIMAGE: ignore noexec /tmp

private-opt standardnotes
```

> Also, like with AppImage, Firejail not detects Standard Notes profile automatically and fallbacks to default one, without explicitly setting it as option. In this package executable is /usr/bin/standard-notes.

Packaging seems to be a bit different depending on OS. But a simple additional redirect can handle this too:

```console
$ cat ~/.config/firejail/standard-notes.profile
# Firejail profile for standard-notes

# Redirect
include standardnotes-desktop.profile

```

With this redirect in place Firejail will pick up the correct profile.

**Reply by hotcapy** [reply]
`ignore noexec /tmp` fixes JavaScript error for both **AppImage** and **AUR** packages. Actually. with it I was able to launch AppImage even without `--no-sandbox` option. So I slightly changed `standardnotes-desktop.local` file (but not sure if it is correct):

```
env LD_PRELOAD=/usr/lib/libhardened_malloc.so
whitelist ${HOME}/.config/lxqt
ignore noexec /tmp
?HAS_APPIMAGE: private-opt standard-notes
```

Still have default cursor theme instead of chosen one, like with some other sandboxed apps. I wonder where LXQt/X11 truly stores it's cursor theme settings. But nothing too important :-)

There are some errors in output related to dbus and pulse, like I mentioned before:

```
[user@host ~]$ standard-notes
Reading profile /home/user/.config/firejail/standard-notes.profile
Reading profile /etc/firejail/standardnotes-desktop.profile
Reading profile /home/user/.config/firejail/standardnotes-desktop.local
Reading profile /etc/firejail/disable-common.inc
Reading profile /etc/firejail/disable-devel.inc
Reading profile /etc/firejail/disable-exec.inc
Reading profile /etc/firejail/disable-interpreters.inc
Reading profile /etc/firejail/disable-programs.inc
Reading profile /etc/firejail/whitelist-var-common.inc
Seccomp list in: !chroot, check list: @default-keep, prelist: unknown,
Parent pid 9275, child pid 9276
Warning: NVIDIA card detected, nogroups command ignored
Warning: An abstract unix socket for session D-BUS might still be available. Use --net or remove unix from --protocol set.
Private /etc installed in 36.64 ms
Warning: NVIDIA card detected, nogroups command ignored
Warning: NVIDIA card detected, nogroups command ignored
Warning: /sbin directory link was not blacklisted
Warning: /usr/sbin directory link was not blacklisted
Warning: not remounting /run/user/1000/doc
Seccomp list in: !chroot, check list: @default-keep, prelist: unknown,
Warning: NVIDIA card detected, nogroups command ignored
Warning: NVIDIA card detected, nogroups command ignored
Child process initialized in 115.06 ms
[31:0327/195106.035073:ERROR:bus.cc(399)] Failed to connect to the bus: Failed to connect to socket /run/firejail/mnt/dbus/system: Permission denied
[31:0327/195106.035218:ERROR:bus.cc(399)] Failed to connect to the bus: Failed to connect to socket /run/firejail/mnt/dbus/system: Permission denied
[31:0327/195106.036605:ERROR:bus.cc(399)] Failed to connect to the bus: Failed to connect to socket /run/firejail/mnt/dbus/user: Permission denied
[31:0327/195106.036727:ERROR:bus.cc(399)] Failed to connect to the bus: Failed to connect to socket /run/firejail/mnt/dbus/user: Permission denied
[31:0327/195106.067620:ERROR:bus.cc(399)] Failed to connect to the bus: Failed to connect to socket /run/firejail/mnt/dbus/user: Permission denied
19:51:06.106 › APPIMAGE env is not defined, current application is not an AppImage
[31:0327/195106.113959:ERROR:bus.cc(399)] Failed to connect to the bus: Failed to connect to socket /run/firejail/mnt/dbus/user: Permission denied
[31:0327/195106.116795:ERROR:bus.cc(399)] Failed to connect to the bus: Failed to connect to socket /run/firejail/mnt/dbus/user: Permission denied
extServer: Server started at http://127.0.0.1:45653
[31:0327/195107.435463:ERROR:bus.cc(399)] Failed to connect to the bus: Failed to connect to socket /run/firejail/mnt/dbus/user: Permission denied
19:51:07.435 › APPIMAGE env is not defined, current application is not an AppImage
[31:0327/195107.438269:ERROR:bus.cc(399)] Failed to connect to the bus: Failed to connect to socket /run/firejail/mnt/dbus/user: Permission denied
Failed to create secure directory (/run/user/1000/pulse): Permission denied
BackupsManager: Data backup successfully saved: 2023-03-27T12-52-22.854Z.txt
```

If app itself runs fine, can I safely ignore them, or can it affect performance/stability in some way?

`include` redirect works great for AUR package.

Thank you for so detailed and helpful answer!

**Comment by kmk3** [comment]
(Offtopic)

> Results are same no matter which directory I choose for placing AppImage:
> `/opt` or `/home/user/Software/Standard-Notes`.

Is ~/Software used by any program or did you create it?

I would recommend placing AppImages in ~/Applications, as firejail makes it
read-only by default.  This is the directory used by [appimaged][1] /
[go-appimaged][2] for downloading/updating AppImages.

Related:

* #5071

[1]: https://github.com/AppImageCommunity/appimaged
[2]: https://github.com/probonopd/go-appimage/blob/master/src/appimaged/README.md

**Reply by hotcapy** [reply]
Oh, `~/Software` is just my own directory, nothing special about it =)

Thank you for useful tip! I will switch to use `~/Applications` for my AppImages.

I'll check appimaged project for sure, didn't try it yet.

**Comment by ghost** [comment]
> So I slightly changed standardnotes-desktop.local file (but not sure if it is correct):
env LD_PRELOAD=/usr/lib/libhardened_malloc.so
whitelist ${HOME}/.config/lxqt
ignore noexec /tmp
?HAS_APPIMAGE: private-opt standard-notes

It looks correct, just one remark. The `?HAS_APPIMAGE: private-opt standard-notes` line translates into `if the command line contains --appimage, add private-opt standard-notes`. The path /opt/standard-notes might need adjusting. For example, in the AUR there are 2 packages offering this 'standard notes' app. Only one installs under /opt but it uses [/opt/standardnotes-bin](https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=standardnotes-bin#n32). If you downloaded the AppImage from its github repo and placed it under /opt/foo you'll need to refer to foo in the private-opt line. Same for any other existing path under /opt that you have this app's executables in. Obviously, if there's nothing related to this app under /opt you can drop the line all together.

> There are some errors in output related to dbus and pulse, like I mentioned before:
If app itself runs fine, can I safely ignore them, or can it affect performance/stability in some way?

You can indeed ignore these errors. They're 'to be expected' due to firejail's options and shouldn't affect performance/stability.

> Still have default cursor theme instead of chosen one, like with some other sandboxed apps. I wonder where LXQt/X11 truly stores it's cursor theme settings.

I don't have an answer to this one. But there are a few things you can try. The first one I would try is adding `include whitelist-common.inc`, because that whitelists ~/.icons and ~/.themes, both of which might hold cursor related files. Alternatively you can go with our [profile template](https://github.com/netblue30/firejail/blob/master/etc/templates/profile.template) as guide and start adding/removing lines untill you find the missing piece of the cursor puzzle. Please do add a note if you happen to find it!

**Reply by hotcapy** [reply]
You're totally right!

Never visited `~/.icons` folder before since thought it stores user apps icons or something similar. I didn't expect that there is actually only a LXQt config file inside with specifically cursor theme!

```
# Written by lxqt-config-appearance
[Icon Theme]
Name=Default
Comment=Default cursor theme
Inherits=Future-cursors
Size=36
```

Despite it mention chosen size, not every app respect it. Standard Notes does, while KeePassXC, which .local profile also has `include whitelist-common.inc` line, doesn't. But anyway, it's awesome to have correct theme, thank you!

**Reply by ghost** [reply]
> Despite it mention chosen size, not every app respect it. Standard Notes does, while KeePassXC, which .local profile also has include whitelist-common.inc line, doesn't. But anyway, it's awesome to have correct theme, thank you!

Ok, great. I'll add `include whitelist-common.inc` to our standardnotes-desktop.profile. Or you can open a PR for it if you'd prefer. Would be nice if you could test if the app still works after a few other includes we usually bring in on whitelisting profiles. This would help fixing the standardnotes-desktop.profile for a future firejail release. So if you find the time, please check with the below options present:

```
include whitelist-common.inc
include whitelist-run-common.inc
include whitelist-runuser-common.inc
include whitelist-usr-share-common.inc
```

**Reply by hotcapy** [reply]
Sure, these options (together with `ignore noexec /tmp`) work fine with both AUR package (**[standardnotes-desktop](https://aur.archlinux.org/packages/standardnotes-desktop)**) and AppImage (manually downloaded, placed to it's folder in `/opt`). I see no additional errors in output or any other kind of problems.

**Reply by ghost** [reply]
Thanks for confirming. I've opened #5768. So remember that a future firejail release will incorporate most if not all of what you have in ~/.config/firejail/standardnotes-desktop.local.

**Comment by kmk3** [comment]
@hotcapy [on Mar 27](https://github.com/netblue30/firejail/discussions/5758#discussioncomment-5441045):

> Still have default cursor theme instead of chosen one, like with some other
> sandboxed apps. I wonder where LXQt/X11 truly stores it's cursor theme
> settings. But nothing too important :-)

LXQt as a DE seems to be relatively modular and self-contained, as it uses
separate programs (such as `lxqt-config-mouse`) to configure different things.

So it might use multiple config paths.

Are there any other LXQt-related paths in your home directory?

This might help finding them:

```sh
cd
find . -maxdepth 5 -iname '*lxqt*'
```

**Reply by hotcapy** [reply]
In my case, just these =)
```
./.config/lxqt
./.config/lxqt/lxqt-runner.conf
./.config/lxqt/lxqt-config-input.conf
./.config/lxqt/lxqt-config-locale.conf
./.config/lxqt/lxqt-config-monitor.conf
./.config/lxqt/lxqt-powermanagement.conf
./.config/lxqt/lxqt-config-file-associations.conf
./.config/lxqt/lxqt.conf
./.config/lxqt/lxqt-config-appearance.conf
./.config/lxqt/lxqt-config.conf
./.config/pcmanfm-qt/lxqt
./.config/lxqt-mimeapps.list
./.cache/pcmanfm-qt/lxqt
./.cache/lxqt-notificationd
```
